### PR TITLE
BDKD-478 Fixed shared volume deletion and list

### PR DESCRIPTION
**Pull request by lolo-fernandez-sirca**

### CODE REVIEW

File: clusterous/cluster.py [hunk 1]
```diff
@@ -1233,26 +1233,27 @@ def terminate_cluster(self, leave_shared_volume, force_delete_shared_volume):
...
-    
-            if leave_shared_volume:
-                self._logger.info('Shared volume "{0}" has not been deleted'.format(shared_volume.id))
-            else:
-                if force_delete_shared_volume or not byo_volume:
-                    if shared_volume.delete():
-                        self._logger.info('Shared volume "{0}" has been deleted'.format(shared_volume.id))
+        attached_volumes = conn.get_all_volumes(filters={'tag:Attached':self.cluster_name})
+        if attached_volumes:
+            for volume in attached_volumes:
```
Comment (balram-ramanathan-sirca) [inline clusterous/cluster.py]: Any reason for this to be a loop? If anything, multiple shared volumes is an error condition.

**Comment by balram-ramanathan-sirca** [comment]
One comment, otherwise looks good

**Comment by lolo-fernandez-sirca** [comment]
@balram-ramanathan-sirca Deleting just the attached shared volume

File: clusterous/cluster.py [hunk 2]
```diff
@@ -1222,9 +1222,23 @@ def terminate_cluster(self, leave_shared_volume, force_delete_shared_volume):
...
         resource_terminated = False
+        instance_list = self._get_instances(self.cluster_name, connection=conn)
+        
+        # Get shared volume ID
+        shared_volume = None
+        byo_volume = False
+        for i in instance_list:
+            if i.tags.get(defaults.instance_node_type_tag_key) == defaults.controller_name_tag_value:
+                volumes = conn.get_all_volumes(filters={'attachment.instance-id': [i.id]})
+                for v in volumes:
```
Comment (balram-ramanathan-sirca) [inline clusterous/cluster.py]: Again, any reason this is a loop? If there is more than one volume here, it is a sign that something has gone wrong, and should arguably result in a warning. The current code will simply loop over all and silently ignore all but the last on the list.

File: clusterous/cluster.py [hunk 3]
```diff
@@ -1222,9 +1222,23 @@ def terminate_cluster(self, leave_shared_volume, force_delete_shared_volume):
...
+                    if v.tags.get('Attached'):
+                        shared_volume = v
+                        byo_volume = True
+
+                    if v.tags.get(defaults.instance_tag_key):
```
Comment (balram-ramanathan-sirca) [inline clusterous/cluster.py]: This can probably be an elif

**Comment by balram-ramanathan-sirca** [comment]
Couple of minor comments. The new logic looks cleaner though.

File: clusterous/cluster.py (hunk 2 above)
Comment (lolo-fernandez-sirca) [inline clusterous/cluster.py]: @balram-ramanathan-sirca It gets only the volumes attached to the "Controller" node. The loop is just to identify which of the two volumes (one is default volume and the other is the attached) is the shared one.

Comment (balram-ramanathan-sirca) [inline clusterous/cluster.py]: That's ok then, can leave as it is.

File: clusterous/cluster.py [hunk 4]
```diff
@@ -1233,26 +1246,24 @@ def terminate_cluster(self, leave_shared_volume, force_delete_shared_volume):
...
                 else:
-                    self._logger.info('Shared volume "{0}" has not been deleted'.format(shared_volume.id))
+                    shared_volume.remove_tags({'Attached': self.cluster_name})
+                    self._logger.info('Leaving shared volume "{0}"'.format(shared_volume.id))
+            else:
+                if leave_shared_volume:
+                    self._logger.info('Leaving shared volume "{0}"'.format(shared_volume.id))
+                else:
+                    if shared_volume.delete():
+                        self._logger.info('Shared volume "{0}" has been deleted'.format(shared_volume.id))
```
Comment (balram-ramanathan-sirca) [inline clusterous/cluster.py]: Probably shouldn't print the shared volume id as default behaviour, as it's inconsistent with the other messages. Perhaps make it debug() or just show "Deleting shared volume", and separately show the shared volume ID with debug().

Comment (lolo-fernandez-sirca) [inline clusterous/cluster.py]: @balram-ramanathan-sirca Good point. Fixed.

**Pull request merged**